### PR TITLE
[Improvement] File Service - Metrics Counter

### PR DESCRIPTION
**Pull request by arjunsk**
## What type of PR is this?
- [ ] API-change
- [ ] BUG
- [X] Improvement
- [ ] Documentation
- [ ] Feature
- [ ] Test and CI
- [ ] Code Refactoring

## Which issue(s) this PR fixes:
- Removing the usage of `trace.span` in File service `mem_cache.go` and `s3_fs.go`. 
- Introducing metrics counter to replace span usage.

issue #

## What this PR does / why we need it:
The below spans were frequently created
- `MemCache.Read` 
- `S3File.Read` 

We are replacing this span with a lighter alternative `metrics.Counter`, which maintains the count instead of span details.

### CODE REVIEW

**Comment by arjunsk** [comment]
The current code is having import cycles. I will be fixing it soon.

File: pkg/util/metric/m_fileservice.go [hunk 1]
```diff
@@ -0,0 +1,29 @@
...
+	false,
+)
+
+var FsS3ReadCounter = NewCounterVec(
+	CounterOpts{
+		Subsystem: "fs",
+		Name:      "s3_read_total",
+		Help:      "Counter for S3 Reads",
+	},
+	[]string{constTenantKey, "type"},
```
Comment (arjunsk) [inline pkg/util/metric/m_fileservice.go:19]: Do we need constTenantKey for `FsS3ReadCounter` & `FsMemCacheReadCounter` ?

File: pkg/util/metric/m_fileservice.go [hunk 2]
```diff
@@ -0,0 +1,29 @@
+package metric
+
+var FsMemCacheReadCounter = NewCounterVec(
+	CounterOpts{
+		Subsystem: "fs",
```
Comment (arjunsk) [inline pkg/util/metric/m_fileservice.go:5]: I assume, Subsystem is "fs"

File: pkg/util/metric/m_fileservice.go [hunk 3]
```diff
@@ -0,0 +1,29 @@
...
+		Name:      "s3_read_total",
+		Help:      "Counter for S3 Reads",
+	},
+	[]string{constTenantKey, "type"},
+	false,
+)
+
+func MemCacheReadCounter() Counter {
+	return FsMemCacheReadCounter.WithLabelValues()
```
Comment (arjunsk) [inline pkg/util/metric/m_fileservice.go:24]: WithLabelValues() is empty.

**Review by reusee — APPROVED** [review]

File: pkg/fileservice/s3_fs.go [hunk 4]
```diff
@@ -372,8 +373,7 @@ func (s *S3FS) Read(ctx context.Context, vector *IOVector) (err error) {
 	default:
 	}
 
-	ctx, span := trace.Start(ctx, "S3FS.Read")
-	defer span.End()
+	metric.S3ReadCounter().Inc()
```
Comment (arjunsk) [inline pkg/fileservice/s3_fs.go:376]: Counter Incr @fengttt

File: pkg/util/metric/m_fileservice.go (hunk 3 above)
Comment (aptend) [reply pkg/util/metric/m_fileservice.go:24]: if the `type` label is not required, you don't have to use `WithLabelValues`, just:
```go
var FsS3ReadCounter = NewCounter(CounterOpts{
	Subsystem: "fs",
	Name:      "s3_read_total",
	Help:      "Counter for S3 Reads",
	ConstLabels: sysTenantID,
})
```
and call FsReadCounter.Inc() directly in fileservice package

Comment (arjunsk) [reply pkg/util/metric/m_fileservice.go]: Thank you. I will update it.

**Review by fengttt — COMMENTED** [review]
I know this is marked WIP.

Metrics still need to register during system initialization.   For example some metrics does not make sense in CN, some does not make sense in DN.   We only want each node to report meaningful metrics.

File: pkg/util/metric/m_register.go [hunk 5]
```diff
@@ -34,6 +34,9 @@ var initCollectors = []Collector{
 	processCollector,
 	// sys metric
 	hardwareStatsCollector,
+	// file service metrics
+	FsMemCacheReadCounter,
+	FsS3ReadCounter,
```
Comment (arjunsk) [inline pkg/util/metric/m_register.go:39]: Metrics defined here are registered during system initialization.

**Comment by arjunsk** [comment]
> Metrics still need to register during system initialization. For example some metrics does not make sense in CN, some does not make sense in DN. We only want each node to report meaningful metrics.

Metrics that are defined in the `m_register.go` are registered during system initialization.
https://github.com/matrixorigin/matrixone/pull/8253/files#r1125619490

File: pkg/fileservice/mem_cache.go [hunk 6]
```diff
@@ -49,8 +48,7 @@ func (m *MemCache) Read(
 ) (
 	err error,
 ) {
-	_, span := trace.Start(ctx, "MemCache.Read")
-	defer span.End()
+	metric.FsMemCacheReadCounter.Inc()
```
Comment (xzxiong) [inline pkg/fileservice/mem_cache.go:51]: as i concerned, MemCacheCount should have 2 type: `hit` & `total`  or  `hit` & `miss`.

File: pkg/fileservice/mem_cache.go [hunk 7]
```diff
@@ -49,8 +48,7 @@ func (m *MemCache) Read(
...
 
 	numHit := 0
 	defer func() {
```
Comment (xzxiong) [inline pkg/fileservice/mem_cache.go:54]: here `stats` already do the count, but no any report output.  is  it better to combine metric with `m.stats` ?

File: pkg/fileservice/mem_cache.go (hunk 6 above)
Comment (arjunsk) [reply pkg/fileservice/mem_cache.go:51]: Sure. I will add the Hit & Miss rates as well. I will refactor this code to be more flexible.

File: pkg/fileservice/mem_cache.go (hunk 7 above)
Comment (arjunsk) [reply pkg/fileservice/mem_cache.go:54]: Yes. That makes sense. I will refactor the code to include these changes.

**Comment by arjunsk** [comment]
After the import cycle fix by @xzxiong , the data is getting persisted in the metrics table

image

Few things, I need to fix

1. Registering `fs` subsystem.
2. Separating `metrics` table into `node_metrics` and `dev_metrics` using flag.
3. Replace stats with counter.Metrics.

File: pkg/fileservice/mem_cache.go (hunk 7 above)
Comment (arjunsk) [reply pkg/fileservice/mem_cache.go:52]: Please find the changes requested in the new PR. https://github.com/matrixorigin/matrixone/pull/8351

File: pkg/fileservice/mem_cache.go (hunk 6 above)
Comment (arjunsk) [reply pkg/fileservice/mem_cache.go]: This PR should address the same: https://github.com/matrixorigin/matrixone/pull/8351

PS: Miss is not explicit in MemCache. We are now exposing NumRead & NumHit. NumMiss is basically NumRead-NumHit.

**Comment by arjunsk** [comment]
Closing this PR as we are now planning to separately handle Dev Metrics and System Metrics. The new PR is : https://github.com/matrixorigin/matrixone/pull/8351

File: pkg/fileservice/mem_cache.go [hunk 8]
```diff
@@ -48,7 +47,6 @@ func (m *MemCache) Read(
 ) (
 	err error,
 ) {
-	metric.FsMemCacheReadCounter.Inc()
```
Comment (arjunsk) [inline pkg/fileservice/mem_cache.go]: Global Metric Defenition. Incr() Usage

**Pull request closed without merging**